### PR TITLE
20230617-builtin-actions

### PR DESCRIPTION
**Pull request by douzzer**
From ChangeLog.md:

Routes can now be configured to match traffic with designated `action_results`
bit constraints, and can be configured to update `action_results` bits, by
inserting the route with a parent event that has the desired configuration.
Parent events can now also be configured to add or clear route flags for all
routes inserted with that parent event.

Added new `aux_event` mechanism to facilitate distinct configurations for a
static generator route and the narrower ephemeral routes dynamically created
when it is matched.

Added a new built-in action, `"%track-peer-v1"`, that can be used in combination
with the above new facilities to dynamically spawn ephemeral routes, allowing
for automatic pinhole routes, automatic adversary tracking, and easy
implementation of dynamic blocks for port scanning adversaries.

For details, see rest of entry in ChangeLog.md for release 1.4.0.

### CODE REVIEW

File: src/addr_families.c [hunk 1]
```diff
@@ -24,8 +24,9 @@
 
 #define WOLFSENTRY_SOURCE_ID WOLFSENTRY_SOURCE_ID_ADDR_FAMILIES_C
 
-static int wolfsentry_addr_family_bynumber_key_cmp(struct wolfsentry_addr_family_bynumber *left, struct wolfsentry_addr_family_bynumber *right) {
-    WOLFSENTRY_RETURN_VALUE(left->number - right->number);
+static int wolfsentry_addr_family_bynumber_key_cmp(const struct wolfsentry_table_ent_header *left, const struct wolfsentry_table_ent_header *right) {
+    WOLFSENTRY_RETURN_VALUE(((const struct wolfsentry_addr_family_bynumber *)left)->number -
```
Comment (JacobBarthelmeh) [inline src/addr_families.c:28]: Tracking down dereferences to number it looks like it is a uint16_t from wolfsentry_settings.h. Is there checks in place that left will 'never' be larger than right?

Looks like its set as a function pointer to .cmp_fn in addr_families.c, didn't track it farther back for where it is called.

File: src/addr_families.c [hunk 2]
```diff
@@ -826,12 +831,75 @@ WOLFSENTRY_API wolfsentry_errcode_t wolfsentry_addr_family_ntop(
 
 #endif /* WOLFSENTRY_PROTOCOL_NAMES */
 
+static wolfsentry_errcode_t wolfsentry_addr_family_max_addr_bits_1(
+    wolfsentry_addr_family_t family,
+    wolfsentry_addr_bits_t *bits)
+{
+    switch(family) {
+    case WOLFSENTRY_AF_INET:
+        { *bits = 32; WOLFSENTRY_RETURN_OK; }
```
Comment (JacobBarthelmeh) [inline src/addr_families.c:840]: What are these max bits for? The IP address?

File: src/json/load_config.c [hunk 3]
```diff
@@ -480,6 +481,62 @@ static wolfsentry_errcode_t convert_default_policy(JSON_TYPE type, const unsigne
 }
 
 static wolfsentry_errcode_t handle_eventconfig_clause(struct wolfsentry_json_process_state *jps, JSON_TYPE type, const unsigned char *data, size_t data_size, struct wolfsentry_eventconfig *eventconfig) {
+    if (jps->cur_depth == 5) {
```
Comment (JacobBarthelmeh) [inline src/json/load_config.c:484]: why at depth 5 for this case?

File: src/routes.c [hunk 4]
```diff
@@ -437,6 +447,92 @@ static wolfsentry_errcode_t wolfsentry_route_init(
...
+    if (((route_exports->flags & WOLFSENTRY_ROUTE_FLAG_SA_FAMILY_WILDCARD) &&
+         ((! (route_exports->flags & WOLFSENTRY_ROUTE_FLAG_SA_REMOTE_ADDR_WILDCARD)) ||
+          (! (route_exports->flags & WOLFSENTRY_ROUTE_FLAG_SA_LOCAL_ADDR_WILDCARD)) ||
+          (! (route_exports->flags & WOLFSENTRY_ROUTE_FLAG_SA_PROTO_WILDCARD)))) ||
+        ((route_exports->flags & WOLFSENTRY_ROUTE_FLAG_SA_PROTO_WILDCARD) &&
+         ((! (route_exports->flags & WOLFSENTRY_ROUTE_FLAG_SA_REMOTE_PORT_WILDCARD)) ||
+          (! (route_exports->flags & WOLFSENTRY_ROUTE_FLAG_SA_LOCAL_PORT_WILDCARD)))))
+        WOLFSENTRY_ERROR_RETURN(INVALID_ARG);
+
+    memset(new,0,offsetof(struct wolfsentry_route, data));
```
Comment (JacobBarthelmeh) [inline src/routes.c:478]: space after commas for readability

File: src/routes.c [hunk 5]
```diff
@@ -437,6 +447,92 @@ static wolfsentry_errcode_t wolfsentry_route_init(
...
+        } else
+            memset(new->data, 0, (size_t)data_addr_offset); /* zero private data. */
+    }
+
+    new->meta.purge_after = route_exports->meta.purge_after;
+    new->meta.connection_count = route_exports->meta.connection_count;
+    new->meta.derogatory_count = route_exports->meta.derogatory_count;
+    new->meta.commendable_count = route_exports->meta.commendable_count;
+
+    if (route_exports->remote.addr_len > 0)
```
Comment (JacobBarthelmeh) [inline src/routes.c:507]: can you adjust this section for readability? Having {} around the if statements when so close to a statement {} on its own, along with limiting it to 80 columns

File: src/addr_families.c (hunk 1 above)
Comment (douzzer) [reply src/addr_families.c:28]: yup that's safe -- `wolfsentry_addr_family_insert()` checks:
```
    if ((max_addr_bits <= 0) || (max_addr_bits > WOLFSENTRY_MAX_ADDR_BITS))
        WOLFSENTRY_ERROR_RETURN(INVALID_ARG);
```
and the struct where it's stored has
```
struct wolfsentry_addr_family_bynumber {
[...]
    wolfsentry_addr_bits_t max_addr_bits;
```
and wolfsentry_settings.h has
```
typedef uint16_t wolfsentry_addr_bits_t;
```
and the limit is set up with asserts to protect it all:
```
#ifndef WOLFSENTRY_MAX_ADDR_BYTES
#define WOLFSENTRY_MAX_ADDR_BYTES 16
#elif WOLFSENTRY_MAX_ADDR_BYTES * 8 > 0xffff
#error WOLFSENTRY_MAX_ADDR_BYTES * 8 must fit in a uint16_t.
#endif

#ifndef WOLFSENTRY_MAX_ADDR_BITS
#define WOLFSENTRY_MAX_ADDR_BITS (WOLFSENTRY_MAX_ADDR_BYTES*8)
#else
#if WOLFSENTRY_MAX_ADDR_BITS > (WOLFSENTRY_MAX_ADDR_BYTES*8)
#error WOLFSENTRY_MAX_ADDR_BITS is too large for given/default WOLFSENTRY_MAX_ADDR_BYTES
#endif
#endif
```

File: src/addr_families.c (hunk 2 above)
Comment (douzzer) [reply src/addr_families.c:840]: yup that's just the usual "a complete IPv4 address is 32 bits".  wolfSentry has to implement some of these sorts of lookups internally for portability.

File: src/json/load_config.c (hunk 3 above)
Comment (douzzer) [reply src/json/load_config.c:484]: that's just where they are in the JSON tree.  the `[` (open-array) pushes it down an extra level.

File: src/routes.c (hunk 4 above)
Comment (douzzer) [reply src/routes.c:478]: fixed, and I found and fixed another one just like it in `wolfsentry_route_init()`.

File: src/routes.c (hunk 5 above)
Comment (douzzer) [reply src/routes.c:507]: good call on the curlies for clarity.  I fixed the line lengths too, but some day I'm going to have to do a full code traversal to fix the rest -- there's 2106 lines over 80 chars wide!

**Pull request merged**